File: harfang/platform/osx/cocoa_window_system.m
// GS Framework - Copyright 2001-2015 Emmanuel Julien. All Rights Reserved.

#import <Cocoa/Cocoa.h>

char cocoa_keyboard[256];
int cocoa_keyboard_flags;

void cocoaInit()
{
	NSLog(@"Cocoa init\n");

	[NSApplication sharedApplication];
	[NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];

	NSLog(@"Cocoa is multithreaded: %d", [NSThread isMultiThreaded]);

	for (int i = 0; i < 256; ++i)
		cocoa_keyboard[i] = 0;
	cocoa_keyboard_flags = 0;
};

NSWindow *cocoaNewWindow(int width, int height, int bpp, int visibility)
{
	NSRect frame = NSMakeRect(0, 0, width, height);
	NSWindow *window = [[NSWindow alloc] initWithContentRect:frame
				styleMask:NSTitledWindowMask | NSResizableWindowMask
				backing:NSBackingStoreBuffered
				defer:NO];

	[window setBackgroundColor:[NSColor blueColor]];

	[window makeKeyAndOrderFront:nil];
	[NSApp activateIgnoringOtherApps:YES];

	NSLog(@"cocoaNewWindow() -> %d", window);
	return window;
}

void ProcessPastEvents()
{
	while (1) {
		NSEvent *event = [NSApp nextEventMatchingMask:NSAnyEventMask
			untilDate:[NSDate distantPast]
			inMode:NSDefaultRunLoopMode
			dequeue:YES];

		if (event == nil)
			break;

		switch (event.type) {
			case NSKeyDown:
				if (event.keyCode < 256)
					cocoa_keyboard[event.keyCode] = 1;
				break;
			case NSKeyUp:
				if (event.keyCode < 256)
					cocoa_keyboard[event.keyCode] = 0;
				break;
			case NSFlagsChanged:
				cocoa_keyboard_flags = [event modifierFlags];
				break;

			case NSMouseMoved:
				break;
			case NSLeftMouseDown:
				break;
			case NSLeftMouseUp:
				break;
			case NSRightMouseDown:
				break;
			case NSRightMouseUp:
				break;
			case NSScrollWheel:
				break;
		}

		[NSApp sendEvent:event];
	}
}

void cocoaUpdateWindow(NSWindow *window) { ProcessPastEvents(); }

void cocoaSetWindowTitle(NSWindow *window, const char *title) { [window setTitle:[NSString stringWithUTF8String:title]]; }

void cocoaDestroyWindow(NSWindow *window)
{
	NSLog(@"cocoaDestroyWindow() -> %d", window);

	[window close];
	[window release];
}

NSView *cocoaGetWindowView(NSWindow *window) { return [window contentView]; }

void cocoaGetViewSize(NSView *view, int *w, int *h)
{
	*w = (int)view.frame.size.width;
	*h = (int)view.frame.size.height;
}

bool CocoaWindowHasFocus(NSWindow *window) { return [window isKeyWindow]; }

void cocoaSetViewSize(NSWindow *window, NSView *view, int w, int h)
{
	NSSize size;
	size.width = w;
	size.height = h + window.frame.size.height - view.frame.size.height;

	NSRect frame = [window frame];
	frame.size = size;
	[window setFrame:frame display:YES animate:NO];
}

int *cocoaGetMonitors(int *count)
{
	NSMutableArray *rects = [NSMutableArray array];
	NSArray *screenArray = [NSScreen screens];
	unsigned screenCount = [screenArray count];
	unsigned index = 0;

	for (index; index < screenCount; index++) {
		NSScreen *screen = [screenArray objectAtIndex:index];
		NSRect screenRect = [screen visibleFrame];
		[rects addObject:[NSNumber numberWithInt:screenRect.origin.x]];
		[rects addObject:[NSNumber numberWithInt:screenRect.origin.y]];
		[rects addObject:[NSNumber numberWithInt:screenRect.size.width]];
		[rects addObject:[NSNumber numberWithInt:screenRect.size.height]];
	}

	*count = [rects count];
	int *cArray = malloc(sizeof(int) * *count);

	for (int i = 0; i < *count; ++i) {
		cArray[i] = [[rects objectAtIndex:i] intValue];
	}

	return cArray;
}

void cocoaDeleteArrayMonitors(int *array) { free(array); }

void cocoaGetWindowPos(NSWindow *window, int *x, int *y)
{
	*x = (int)window.frame.origin.x;
	*y = (int)window.frame.origin.y + window.frame.size.height;
}

void cocoaSetWindowPos(NSWindow *window, int x, int y)
{
	NSPoint pos;
	pos.x = x;
	pos.y = y;
	[window setFrameTopLeftPoint:pos];
}

File: harfang/platform/osx/cocoa_window_system.cpp
// HARFANG(R) Copyright (C) 2021 Emmanuel Julien, NWNC HARFANG. Released under GPL/LGPL/Commercial Licence, see licence.txt for details.

#include "window_system/window_system.h"
#include "cocoa_window_system/cocoa_window_system.h"
#include "cstl/main_thread.h"

namespace gs {

struct MonitorCocoa : Monitor {
	MonitorCocoa(string _id) : Monitor() { id = _id; }
	iRect GetRect() override { return rect; }
	iRect rect;
};

static const char *window_type = "Cocoa";

struct WindowCocoa : Window
{
	WindowCocoa() : window(nullptr), view(nullptr) { type = window_type; }

	NSWindow *window;
	NSView *view;
};

#define ASSERT_WINDOW_TYPE(W, R) \
	if ((W) == nullptr || (W)->type != window_type) \
		return (R);

#define CAST_WINDOW_SAFE(W, R) \
	ASSERT_WINDOW_TYPE(W, R)   \
	auto w_cocoa = reinterpret_cast<WindowCocoa *>(W);

#define CAST_CONST_WINDOW_SAFE(W, R) \
	ASSERT_WINDOW_TYPE(W, R)         \
	auto w_cocoa = reinterpret_cast<const WindowCocoa *>(W);

//
void WindowSystemInit() { cocoaInit(); }

Window *NewWindow(int width, int height, int bpp, Window::Visibility visibility)
{
	scoped_ptr<WindowCocoa> w(new WindowCocoa);

	switch (visibility) {
		case Window::Fullscreen: {
		} break;

		case Window::Undecorated:
			break;

		default:
		case Window::Windowed:
			w->window = main_thread::queue(bind(&cocoaNewWindow, width, height, bpp, 2)).get();
			break;
	}

	w->view = cocoaGetWindowView(w->window);

	SetWindowTitle(w, "Harfang");

	new_window_signal.Emit(w);
	return w.detach();
}

Window *NewWindowFrom(void *handle)
{
	auto w = new WindowCocoa;

	w->is_foreign = true;
	w->view = reinterpret_cast<NSView *>(handle);

	return w;
}

Window::Handle GetWindowHandle(const Window *w)
{
	CAST_CONST_WINDOW_SAFE(w, nullptr);
	return w_cocoa->view;
}

bool DestroyWindow(Window *w)
{
	CAST_WINDOW_SAFE(w, false);
	if (w_cocoa->window != nullptr)
		main_thread::queue(bind(&cocoaDestroyWindow, w_cocoa->window)).wait();
	return true;
}

void DeleteWindow(Window *w) { delete w; }

bool UpdateWindow(Window *w)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	cocoaUpdateWindow(w_cocoa->window);
	return true;
}

//
bool GetWindowClientSize(const Window *w, int &width, int &height)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	cocoaGetViewSize(w_cocoa->view, &width, &height);
	return true;
}

//
bool SetWindowClientSize(const Window *w, int width, int height)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	cocoaSetViewSize(w_cocoa->window, w_cocoa->view, width, height);
	return true;
}

bool GetWindowTitle(const Window *w, string &title)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	return false;
}

bool SetWindowTitle(Window *w, const string &title)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	cocoaSetWindowTitle(w_cocoa->window, title.to_utf8());
	return true;
}

void GetMonitors(shared_vector<Monitor *> &monitor_array)
{
	int count = 0;
	int *rects = cocoaGetMonitors(&count);
	for (int i = 0; i < count; i += 4)
	{
		shared_ptr<MonitorCocoa> m(new MonitorCocoa(""));
		m->rect.Set(rects[i], rects[i + 1], rects[i + 2], rects[i + 3]);
		monitor_array.push_back(m);
	}
	cocoaDeleteArrayMonitors(rects);
}

iVector2 GetWindowPos(Window *w)
{
	CAST_CONST_WINDOW_SAFE(w, iVector2());
	iVector2 pos;
	cocoaGetWindowPos(w_cocoa->window, &pos.x, &pos.y);
	return pos;
}

bool SetWindowPos(Window *w, const iVector2 v)
{
	CAST_WINDOW_SAFE(w, false);
	cocoaSetWindowPos(w_cocoa->window, v.x, v.y);
	return true;
}

bool WindowHasFocus(const Window *w)
{
	CAST_CONST_WINDOW_SAFE(w, false);
	return CocoaWindowHasFocus(w_cocoa->window);
}

// TODO
void ShowCursor() {}
void HideCursor() {}

} // hg
